format.go

```diff
@@ -0,0 +1,204 @@
+// Copyright 2015 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Adapted from math/big/ftoa.go.
+
+package apd
+
+import (
+	"fmt"
+	"strconv"
+)
+
+// Text converts the floating-point number x to a string according
+// to the given format. The format is one of:
+//
+//	'e'	-d.dddde±dd, decimal exponent, exponent digits
+//	'E'	-d.ddddE±dd, decimal exponent, exponent digits
+//	'f'	-ddddd.dddd, no exponent
+//	'g'	like 'e' for large exponents, like 'f' otherwise
+//	'G'	like 'E' for large exponents, like 'f' otherwise
+//
+// If format is a different character, Text returns a "%" followed by the
+// unrecognized.Format character. The 'f' format has the possibility of
+// displaying precision that is not present in the Decimal when it appends
+// zeros. All other formats always show the exact precision of the Decimal.
+func (d *Decimal) Text(format byte) string {
+	cap := 10 // TODO(gri) determine a good/better value here
+	return string(d.Append(make([]byte, 0, cap), format))
+}
+
+// String formats x like x.Text('G'). It matches the to-scientific-string
+// conversion of the GDA spec.
+func (d *Decimal) String() string {
+	return d.Text('G')
+}
+
+// Append appends to buf the string form of the decimal number d,
+// as generated by d.Text, and returns the extended buffer.
+func (d *Decimal) Append(buf []byte, fmt byte) []byte {
+	// sign
+	if d.Negative {
+		buf = append(buf, '-')
+	}
+
+	switch d.Form {
+	case Finite:
+		// ignore
+	case NaN:
+		return append(buf, "NaN"...)
+	case NaNSignaling:
+		return append(buf, "sNaN"...)
+	case Infinite:
+		return append(buf, "Infinity"...)
+	default:
+		return append(buf, "unknown"...)
+	}
+
+	digits := d.Coeff.String()
+	switch fmt {
+	case 'e', 'E':
+		return fmtE(buf, fmt, d, digits)
+	case 'f':
+		return fmtF(buf, d, digits)
+	case 'g', 'G':
+		// See: http://speleotrove.com/decimal/daconvs.html#reftostr
+		const adjExponentLimit = -6
+		adj := int(d.Exponent) + (len(digits) - 1)
+		if d.Exponent <= 0 && adj >= adjExponentLimit {
+			return fmtF(buf, d, digits)
+		}
+		return fmtE(buf, fmt+'e'-'g', d, digits)
```

Comment:
This is copied from https://github.com/golang/go/blob/fcc35147d5940d5b077586c02a97894c3edf870d/src/math/big/ftoa.go?utf8=%E2%9C%93#L148

It needs to convert the either `g` or `G` into a `e` or `E` since that's what `fmtE` expects. So it is indeed `fmt - 2`, but attempting to do that in a way that illustrates the intention.